runner/execer/os/os_execer.go

```diff
@@ -395,29 +388,141 @@ func (p *osProcess) Wait() (result execer.ProcessStatus) {
 	return result
 }
 
-func (p *osProcess) Abort() (result execer.ProcessStatus) {
+// Attempt to SIGTERM process, allowing for graceful exit
+// SIGKILL after 10 seconds or if osProcess.cmd.Wait() returns an error
+func (p *osProcess) Abort() execer.ProcessStatus {
 	p.mutex.Lock()
 	defer p.mutex.Unlock()
 	if p.result != nil {
 		return *p.result
 	} else {
-		p.result = &result
+		p.result = &execer.ProcessStatus{}
 	}
-	result.State = execer.FAILED
-	result.ExitCode = -1
-	result.Error = "Aborted."
+	p.result.State = execer.FAILED
+	p.result.ExitCode = -1
+	p.result.Error = "Aborted."
+
+	sigCh := make(chan error)
+	var err error
+
+	go func() {
+		sigCh <- p.cmd.Process.Signal(syscall.SIGTERM)
+	}()
+
+	select {
+	case err = <-sigCh:
+		if err != nil {
+			msg := fmt.Sprintf("Error aborting command via SIGTERM: %s.", err)
+			log.WithFields(
+				log.Fields{
+					"pid":    p.cmd.Process.Pid,
+					"tag":    p.Tag,
+					"jobID":  p.JobID,
+					"taskID": p.TaskID,
+				}).Errorf(msg)
+			p.KillAndWait(msg)
+		} else {
+			log.WithFields(
+				log.Fields{
+					"pid":    p.cmd.Process.Pid,
+					"tag":    p.Tag,
+					"jobID":  p.JobID,
+					"taskID": p.TaskID,
+				}).Info("Aborting process via SIGTERM")
+		}
+	}
+
+	// Add buffer in case of race condition where both <-cmdDoneCh returns an error & timeout is exceeded at same time
+	errCh := make(chan error, 1)
+	go func() {
+		select {
+		case <-time.After(time.Second * 10):
+			errCh <- errors.New("10 second timeout exceeded.")
+		}
+	}()
+
+	cmdDoneCh := make(chan error)
+	go func() {
+		cmdDoneCh <- p.cmd.Wait()
+	}()
+
+	for {
+		select {
+		case err = <-cmdDoneCh:
+			if err != nil {
+				msg := fmt.Sprintf("Command failed to finish: %v", err)
```

Comment:
"failed to finish [+successfully]" - is that more 'correct'?

Reply:
sgtm